### PR TITLE
Remove whitespace generation

### PR DESCRIPTION
**Pull request by danez**
| Q                        | A 
| ------------------------ | ---
| Patch: Bug Fix?          | n
| Major: Breaking Change?  | y
| Minor: New Feature?      | n
| Deprecations?            | n
| Spec Compliancy?         | n
| Tests Added/Pass?        | y
| Fixed Tickets            | 
| License                  | MIT
| Doc PR                   | 
| Dependency Changes       | 

This removes the whitespace generation and the generator would now simply print everything  according to the default settings and not based on the AST tokens.
By removing this babel can use babylon without tokens, which brings a performance boost in babylon.

This change on it's own doesn't change the performance much. I tried printing react a 100 times with (14.8s) and without (15.0s) this change and it is neglectable, but at least not bad.

I removed all expected files and regenerated them.

### CODE REVIEW

File: packages/babel-generator/test/fixtures/comments/computed-property-comments-2/expected.js [hunk 1]
```diff
@@ -1,11 +1,9 @@
 var test = {
   /**
    * Before bracket init
-   */
-  ["a"]: "1",
+   */["a"]: "1",
```
Comment (loganfsmyth) [inline packages/babel-generator/test/fixtures/comments/computed-property-comments-2/expected.js]: Seems like a bad case here.

File: packages/babel-generator/test/fixtures/comments/simple-a-lot-of-line-comment/expected.js [hunk 2]
```diff
@@ -19,11 +19,7 @@
 // ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 // (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 // THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-
-
-function test() {}
-
-// Copyright (C) 2012 Yusuke Suzuki <utatane.tea@gmail.com>
+function test() {} // Copyright (C) 2012 Yusuke Suzuki <utatane.tea@gmail.com>
```
Comment (loganfsmyth) [inline packages/babel-generator/test/fixtures/comments/simple-a-lot-of-line-comment/expected.js]: Hmm, also an unfortunate case.

**Comment by loganfsmyth** [comment]
Love this PR, anything to simplify codegen is a plus in my book.

I'd vote for including newlines after all block comments to make some of these examples nicer, and maybe to include a newline after all declaration statements?

**Comment by danez** [comment]
Yes will look into this, what I also noticed is that 
```js
function a () {
};
```
gets reprinted as
```js
function a() {
}

;
```
But this is because of the `EmptyStatement`. I wasn't sure how to fix that because the generation always prints a newline after a Node. And I didn't want to create special hacks in the code for `EmptyStatement`.

**Comment by danez** [comment]
Okay I change the generation now:

* Add newline after last empty SwitchCase, to avoid 

```
switch {
case 'a':
  break;
default:}
```

* Add newlines around block comments if they are non-flow comments or contain newlines

I also made two commits one with the code changes and one with the test fixes.

File: packages/babel-generator/test/fixtures/flow/call-properties/expected.js [hunk 3]
```diff
@@ -1,5 +1,14 @@
-var a: { (): number };
-var a: { (): number };
-var a: { y: string, (): number, (x: string): string, };
-var a: { <T>(x: T): number };
-interface A { (): number };
+var a: { (): number
```
Comment (loganfsmyth) [inline packages/babel-generator/test/fixtures/flow/call-properties/expected.js]: Weird asymmetry on these objects. I think if the closing bracket is on its own line, the opening one should be too.

File: packages/babel-generator/test/fixtures/flow/interfaces-module-and-script/expected.js [hunk 4]
```diff
@@ -1,9 +1,23 @@
-interface A {};
-interface A extends B {};
-interface A<T> extends B<T>, C<T> {};
-interface A { foo: () => number };
-interface Dictionary { length: number, [index: string]: string, };
+interface A {}
+;
```
Comment (loganfsmyth) [inline packages/babel-generator/test/fixtures/flow/interfaces-module-and-script/expected.js]: Are these actually parsed as `EmptyStatement` nodes, or is the semicolon being printed weirdly.

File: packages/babel-generator/src/printer.js [hunk 5]
```diff
@@ -538,8 +513,17 @@ export default class Printer {
...
     }
 
-    // whitespace before
-    this.newline(this._whitespace ? this._whitespace.getNewlinesBefore(comment) : 0);
+    // We leave inline flow block comments inline
+    const isNonFlowBlockComment =
+      comment.type === "CommentBlock" &&
+      (
+        (!comment.value.startsWith(":") &&
+        !comment.value.startsWith("flow-include")) ||
```
Comment (loganfsmyth) [inline packages/babel-generator/src/printer.js]: Do we have any tests for this? Does Flow actually require these to be on the same line, or is this just for making it nicer to read? I don't love having a special case like this.

File: packages/babel-generator/test/fixtures/flow/call-properties/expected.js (hunk 3 above)
Comment (danez) [inline packages/babel-generator/test/fixtures/flow/call-properties/expected.js]: Fixed

File: packages/babel-generator/test/fixtures/flow/interfaces-module-and-script/expected.js (hunk 4 above)
Comment (danez) [inline packages/babel-generator/test/fixtures/flow/interfaces-module-and-script/expected.js]: Not sure about this and how to approach. There is always an EmptyStatement generated after declarations with semicolon, and I'm not sure how we could ignore it if possible.

File: packages/babel-generator/src/printer.js (hunk 5 above)
Comment (danez) [inline packages/babel-generator/src/printer.js]: I just tested and they don't need to be on the same line. Will remove

Comment (danez) [inline packages/babel-generator/src/printer.js]: Should I leave the case that only block comments with newlines inside them get newlines around them? Or add newlines around to all block comments?

Comment (danez) [inline packages/babel-generator/src/printer.js]: I removed it now, all BlockComments have newline around them now.

File: packages/babel-generator/test/fixtures/flow/interfaces-module-and-script/expected.js (hunk 4 above)
Comment (loganfsmyth) [inline packages/babel-generator/test/fixtures/flow/interfaces-module-and-script/expected.js]: Looking at the AST from Flow itself, the semicolons are not in the grammar for interface declarations, which means the current behavior of parsing them as EmptyStatement nodes is correct. We should probably just remove the semicolons from the `actual.js` files, but I understand if you don't want to for now.

**Comment by hzoo** [comment]
Are we removing end of file newlines now?

**Comment by danez** [comment]
Not sure, in the fixture runner they are always trimmed. I will check

**Comment by danez** [comment]
@hzoo It seems to me we are always removing whitespaces at the end already here: https://github.com/babel/babel/blob/7.0/packages/babel-generator/src/buffer.js#L45

**Comment by peey** [comment]
Could we do this - generate code but don't worry too much about styling and then pipe it through a linter's autofix for style consistency in the generated code?

**Comment by hzoo** [comment]
> then pipe it through a linter's autofix for style consistency in the generated code

The purpose was to make babel-generator faster so having to use another tool would go against this idea especially since it's easier to just encode the logic in the printer itself rather than in an autofixer. And if we were going to do that I would just use prettier as the printer anyway (which is much slower than hardcoding something simple)?

I think we just want it to look presentable but focus on speed. Again it's generated output not the source code (why we removed the options for quotes).

**Comment by danez** [comment]
Let me know if I should change more in this PR.

File: packages/babel-generator/src/node/whitespace.js [hunk 6]
```diff
@@ -80,9 +80,10 @@ export const nodes = {
    * Test if SwitchCase needs whitespace.
    */
 
-  SwitchCase(node: Object, parent: Object): ?WhitespaceObject {
+  SwitchCase(node: Object, parent: Object): WhitespaceObject {
     return {
       before: node.consequent.length || parent.cases[0] === node,
+      after: !node.consequent.length && parent.cases[parent.cases.length - 1] === node,
```
Comment (jridgewell) [inline packages/babel-generator/src/node/whitespace.js]: This doesn't make sense.

File: packages/babel-generator/src/node/whitespace.js [hunk 7]
```diff
@@ -174,6 +175,29 @@ nodes.ObjectMethod = function (node: Object, parent): ?WhitespaceObject {
   }
 };
 
+nodes.ObjectTypeCallProperty = function (node: Object, parent): ?WhitespaceObject {
+  if (
+    parent.callProperties[0] === node &&
+    (!parent.properties || !parent.properties.length)
+    ) {
```
Comment (jridgewell) [inline packages/babel-generator/src/node/whitespace.js]: Negative indent.

File: packages/babel-generator/src/node/whitespace.js (hunk 6 above)
Comment (danez) [inline packages/babel-generator/src/node/whitespace.js]: Why not? It only covers the case of empty last case

```
switch() {
  default:
}
```

Comment (jridgewell) [inline packages/babel-generator/src/node/whitespace.js]: Oh, i read this as two `!node.consequent.length && last(node.consequent)`. Never mind.

**Comment by hzoo** [comment]
Alright then!

**Comment by danez** [comment]
Thanks for taking care.

**Comment by kittens** [comment]
For what it's worth, you can still retain newlines when printing a statement list by comparing the `loc.end.line` and `loc.start.line` of two adjacent statements and inserting the correct amount of newlines.

**Pull request merged**